generator/json-generator.lua

```diff
@@ -1,135 +1,162 @@
+    return { type = 'ref', name = type }
 end
 
-local function cleanUpString( str )
-    return str:gsub( '\n\n', ' ' ):gsub( '\"', '\\"' ):gsub( '\0', '' )
+local function createReturnTypes( returns )
+    local returnTypes = {}
+    for i, v in ipairs( returns ) do
+        returnTypes[i] = getType( v.type )
+    end
+    return returnTypes
 end
 
-local function buildCallbackCompletion( f )
-    local arguments = generateArguments( f.variants[1].arguments )
-    return {
-        TAB .. '{' .. LINE_BREAK,
-        TAB .. TAB .. '"displayText": "love.' .. f.name .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"type": "function"' .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"description": ' .. APOSTROPHE .. cleanUpString( f.description ) .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"descriptionMoreURL": ' .. APOSTROPHE .. WIKI_URL .. 'love.' .. f.name .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"snippet": ',
-        APOSTROPHE .. 'love.' .. string.format( f.name .. '(%s)', arguments ) .. APOSTROPHE .. LINE_BREAK,
-        TAB .. '},' .. LINE_BREAK
-    }
+-- TODO handle functions with neither args nor return variables.
+local function createVariant( vdef )
+    local variant = {}
+    if vdef.arguments then
+        variant.args = createArguments( vdef.arguments )
+    end
+
+    if vdef.returns then
+        variant.returnTypes = createReturnTypes( vdef.returns )
+    end
+    return variant
 end
 
-local function buildModuleFunctionCompletion( f, module, closing )
-    local arguments = generateArguments( f.variants[1].arguments )
-    return {
-        TAB .. '{' .. LINE_BREAK,
-        TAB .. TAB .. '"displayText": "love.' .. module.name .. '.' .. f.name .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"type": "function"' .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"description": ' .. APOSTROPHE .. cleanUpString( f.description ) .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"descriptionMoreURL": ' .. APOSTROPHE .. WIKI_URL .. string.format( 'love.%s.%s', module.name, f.name ) .. APOSTROPHE .. COMMA .. LINE_BREAK,
-        TAB .. TAB .. '"snippet": ',
-        APOSTROPHE .. string.format( 'love.%s.%s(%s)', module.name, f.name, arguments ) .. APOSTROPHE .. LINE_BREAK,
-        TAB .. (  closing and '}' or '},'  ) .. LINE_BREAK
-    }
+local function createFunction( f, parent, moduleString )
+    local name = f.name
+    parent[name] = {}
+    parent[name].type = 'function'
+
+    parent[name].description = f.description
+    parent[name].link = string.format( "%s%s%s", WIKI_URL, moduleString, name )
+
+    -- Create normal function if there is just one variant.
+    if #f.variants == 1 then
+        local v = createVariant( f.variants[1] )
+        parent[name].args = v.args
+        parent[name].returnTypes = v.returnTypes
+        return
+    end
+
+    -- Generate multiple variants.
+    parent[name].variants = {}
+    for i, v in ipairs( f.variants ) do
```

Comment:
There is a `return` in the if statement above the code you marked. So unless I'm misunderstanding something here the code shouldn't execute.